### PR TITLE
Optimize performance of `math::trunc` (~2.5x faster)

### PR DESCRIPTION
**Pull request by tlm365**
## Rationale for this change
Same idea as https://github.com/apache/datafusion/pull/12881. Using the `unary`/`binary` functions allow faster processing (most likely auto-vectorized code) by avoiding branching on nulls.

## What changes are included in this PR?
- Apply `unary` and `binary`
- Add benchmark

## Are these changes tested?
Existing testcases.

## Are there any user-facing changes?
No.

**BENCHMARK RESULT
screenshot_2024-10-13_19-49-09

### CODE REVIEW

File: datafusion/functions/src/math/trunc.rs [hunk 1]
```diff
@@ -111,44 +111,66 @@ fn trunc(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                                x.trunc()
+                            }
+                        }),
+                ) as ArrayRef)
+            }
+            ColumnarValue::Array(precision) => {
+                let num_array = num.as_primitive::<Float32Type>();
+                let precision_array = precision.as_primitive::<Int64Type>();
+                let result: PrimitiveArray<Float32Type> =
+                    arrow_arith::arity::binary(num_array, precision_array, |x, y| {
```
Comment (simonvandel) [inline datafusion/functions/src/math/trunc.rs:168]: Is this the same function as this? https://docs.rs/arrow/latest/arrow/compute/fn.binary.html

If so you could remove the new dependency (although it is in the dependency closure anyway already)

Comment (tlm365) [reply datafusion/functions/src/math/trunc.rs:168]: @simonvandel Thanks so much for reviewing
> Is this the same function as this? https://docs.rs/arrow/latest/arrow/compute/fn.binary.html

Oh, yes it is. Nice!

**Review by alamb — APPROVED** [review]
Thanks @tlm365 and @simonvandel

File: datafusion/functions/src/math/trunc.rs [hunk 2]
```diff
@@ -111,44 +111,66 @@ fn trunc(args: &[ArrayRef]) -> Result<ArrayRef> {
...
         ColumnarValue::Array(Arc::clone(&args[1]))
     };
 
-    match args[0].data_type() {
+    match num.data_type() {
         Float64 => match precision {
-            ColumnarValue::Scalar(Int64(Some(0))) => Ok(Arc::new(
-                make_function_scalar_inputs!(num, "num", Float64Array, { f64::trunc }),
-            ) as ArrayRef),
-            ColumnarValue::Array(precision) => Ok(Arc::new(make_function_inputs2!(
```
Comment (findepi) [inline datafusion/functions/src/math/trunc.rs:128]: I've broadened https://github.com/apache/datafusion/issues/12923 to cover this too

**Comment by alamb** [comment]
I merged up from main to resolve a conflict on this branch

**Comment by alamb** [comment]
🚀  thanks again

**Pull request merged**